packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -139,48 +146,85 @@ const ElementHook: Hooks<VElement> = {
     },
 };
 
+function isUpgradableComponent(
+    ctor: CustomElementConstructor | UpgradableCustomElementConstructor
+): ctor is UpgradableCustomElementConstructor {
+    return 'upgrade' in ctor;
+}
+
 const CustomElementHook: Hooks<VCustomElement> = {
     create: (vnode) => {
         const {
             sel,
             owner: { renderer },
         } = vnode;
-        const elm = renderer.createElement(sel);
-
-        linkNodeToShadow(elm, vnode);
-        createViewModelHook(elm, vnode);
+        const UpgradableConstructor = getUpgradableConstructor(sel, renderer);
+        if (isUpgradableComponent(UpgradableConstructor)) {
+            // the custom element from the registry is expecting an upgrade callback
+            UpgradableConstructor.upgrade = (elm: HTMLElement) => {
+                createViewModelHook(elm, vnode);
+            };
+        }
+        const elm = new UpgradableConstructor();
         vnode.elm = elm;
+        linkNodeToShadow(elm, vnode);
 
-        allocateChildrenHook(vnode);
+        const vm = getAssociatedVMIfPresent(elm);
+        if (vm) {
+            allocateChildrenHook(vnode, vm);
+        } else if (vnode.ctor !== UpgradableConstructor) {
+            throw new TypeError(`Incorrect Component Constructor`);
```

Comment:
@ravijayaramappa @pmdartus this is the line to throw if the resolved constructor for a CE does not match the one used on the upgrade process to protect ourself until scoped registry lands.